File: web-scraper-node/scrape.js
import puppeteer from 'puppeteer';
import fs from 'fs';

const scrape = async () => {
  const browser = await puppeteer.launch();
  const page = await browser.newPage();

  const allBooks = [];
  let currentPage = 1;
  const maxPages = 10;

  while (currentPage <= maxPages) {
    const url = `https://books.toscrape.com/catalogue/page-${currentPage}.html`;

    await page.goto(url);

    const books = await page.evaluate(() => {
      const bookElements = document.querySelectorAll('.product_pod');
      return Array.from(bookElements).map((book) => {
        const title = book.querySelector('h3 a').getAttribute('title');
        const price = book.querySelector('.price_color').textContent;
        const stock = book.querySelector('.instock.availability')
          ? 'In Stock'
          : 'Out Of Stock';
        const rating = book
          .querySelector('.star-rating')
          .className.split(' ')[1];
        const link = book.querySelector('h3 a').getAttribute('href');

        return {
          title,
          price,
          stock,
          rating,
          link,
        };
      });
    });

    allBooks.push(...books);
    console.log(`Books on page ${currentPage}: `, books);
    currentPage++;
  }

  fs.writeFileSync('books.json', JSON.stringify(allBooks, null, 2));

  console.log('Data saved to books.json');

  await browser.close();
};

scrape();

File: web-scraper-python/scrape.py
import requests
from bs4 import BeautifulSoup
import json

def fetch_books(page_number):
  url = f"https://books.toscrape.com/catalogue/page-{page_number}.html"
  response = requests.get(url)
  soup = BeautifulSoup(response.text, 'html.parser')

  books = []
  book_elements = soup.find_all('article', class_='product_pod')

  for book in book_elements:
    title = book.find('h3').find('a')['title']
    price = book.find('p', class_='price_color').text
    stock = 'In stock' if 'In stock' in book.find('p', class_='instock availability').text else 'Out of stock'
    rating = book.find('p', class_='star-rating')['class'][1]
    link = book.find('h3').find('a')['href']

    books.append({
      'title': title,
      'price': price,
      'stock': stock,
      'rating': rating,
      'link': f"https://books.toscrape.com/catalogue/{link}"
    })

  return books

def main():
  all_books = []
  max_pages = 10

  for current_page in range(1, max_pages + 1):
    books_on_page = fetch_books(current_page)
    all_books.extend(books_on_page)
    print(f"Books on page {current_page}: {books_on_page}")

  # Save data to file
  with open('books.json', 'w') as f:
    json.dump(all_books, f, indent=2)

  print('Data is saved to books.json')

if __name__ == "__main__":
  main()